plugin-java/src/main/java/appland/javaAgent/AppMapJavaAgentDownloadService.java

```diff
@@ -0,0 +1,173 @@
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+/**
+ * Application service to manage the download of the AppMap Java agent.
+ */
+public final class AppMapJavaAgentDownloadService {
+    private static final Logger LOG = Logger.getInstance(AppMapJavaAgentDownloadService.class);
+    private static final String AGENT_LINK_FILENAME = "appmap.jar";
+
+    public static AppMapJavaAgentDownloadService getInstance() {
+        return ApplicationManager.getApplication().getService(AppMapJavaAgentDownloadService.class);
+    }
+
+    /**
+     * @return Local path to the downloaded Java agent file.
+     * {@code null} is returned if the agent has not been downloaded yet.
+     */
+    public @Nullable Path getJavaAgentPathIfExists() {
+        var jarFilePath = getJavaAgentPath();
+        return jarFilePath == null || Files.notExists(jarFilePath) ? null : jarFilePath;
+    }
+
+    @SuppressWarnings("DialogTitleCapitalization")
+    public void downloadJavaAgent(@NotNull Project project) {
+        var title = AppMapBundle.get("javaAgent.download.title");
+        new Task.Backgroundable(project, title, false, PerformInBackgroundOption.ALWAYS_BACKGROUND) {
+            @Override
+            public void run(@NotNull ProgressIndicator indicator) {
+                try {
+                    downloadJavaAgentSync(indicator);
+                } catch (IOException e) {
+                    LOG.warn("failed to download AppMap Java agent", e);
+                }
+            }
+        }.queue();
+    }
+
+    public boolean downloadJavaAgentSync(@NotNull ProgressIndicator indicator) throws IOException {
+        var agentFilePath = getJavaAgentPath();
+        if (agentFilePath == null) {
+            LOG.warn("unable to locate target file path for AppMap Java agent");
+            return false;
+        }
+
+        var latestAsset = GitHubRelease.getLatestRelease(indicator, "getappmap", "appmap-java")
+                .stream()
+                .filter(asset -> "application/java-archive".equals(asset.getContentType()))
+                .findFirst()
+                .orElse(null);
+
+        if (latestAsset == null) {
+            LOG.warn("JAR assets not found in latest GitHub release of getappmap/appmap-java");
+            return false;
+        }
+
+        // e.g. ~/.appmap/lib/java/appmap-1.7.2.jar
+        var assetDownloadPath = agentFilePath.resolveSibling(latestAsset.getFileName());
+
+        // don't download if the asset file does already exist
+        if (Files.exists(assetDownloadPath)) {
+            return false;
+        }
+
+        // e.g. appmap-1.7.2.downloading
+        var lockFilePath = assetDownloadPath.resolveSibling(latestAsset.getFileName().replace(".jar", ".downloading"));
```

Comment:
The lockfile is `appmap-X.X.X.jar.downloading`, not `appmap-X.X.X.downloading`, so you don't want to use the `replace` function. You can just add `.downloading` to the end of the file name.

Reply:
Fixed by https://github.com/getappmap/appmap-intellij-plugin/pull/300/commits/0ead3e9c961bdf0a97243753352e42839d334b04